### PR TITLE
add blogging section using Contentlayer🚀

### LINKED ISSUE #59: [OPE-52] Create ContentLayer Blog
<sub>From [SyncLinear.com](https://synclinear.com) | [OPE-52](https://linear.app/openstatus/issue/OPE-52/create-contentlayer-blog)</sub>

### PR DESCRIPTION
**Pull request by nkilm**
## Major Changes

- New blog page🎉
- All blogs are stored inside `src/content/blog` as suggested by @mxkaske 
- All routes for blogs are public - `/blog/*`

## Blog format
```md
---
title: Hello World
description: Blog using Contentlayer
author: Nikhil Mohite
authorLink: https://nikhilmohite.com
publishedAt: 2023-07-18
---

Hello World!
```

## Fixes

- #59

## Possible improvements/Additions

- The UI can be revamped to make the blog more visually appealing.
- Adding `Table of Contents` to each blog.
- New MDX components.
- Cover image for each blog - can be used as [OG](https://ogp.me/) image also.
- Add `copy` button for each code snippet.

## Known issues with [contentlayer](https://contentlayer.dev)

- https://github.com/contentlayerdev/contentlayer/issues/506

For now, the solution provided is to override the package versions in `package.json`

--- 

## Visuals

image

image

#### Coding snippets

Theme used - `one-dark-pro`

image

### CODE REVIEW

File: apps/web/src/app/_components/background.tsx [hunk 1]
```diff
@@ -12,7 +13,7 @@ export default function Background({
   useMouseMove();
   return (
     <>
-      <div className="fixed h-full w-full">
+      <div className="fixed -z-50 h-full w-full">
```
Comment (mxkaske) [inline apps/web/src/app/_components/background.tsx:16]: Good catch!

File: apps/web/src/app/blog/[slug]/page.tsx [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import { Mdx } from "@/components/Mdx";
+import { Button } from "@/components/ui/button";
+
+export async function generateStaticParams() {
+  return allBlogs.map((blog) => ({
+    slug: blog.slug,
+  }));
+}
+
+// TODO: add OG image
```
Comment (mxkaske) [inline apps/web/src/app/blog/[slug]/page.tsx]: For the OG image, you can use following link that will automatically generate the `title` and `description` inside of it: `https://openstatus.dev/api/og?title=${title}&description=${description}`.

File: apps/web/src/app/blog/[slug]/page.tsx [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+export default function BlogPage({ params }: { params: { slug: string } }) {
+  const blog = allBlogs.find((blog) => blog.slug === params.slug);
+
+  if (!blog) {
+    notFound();
+  }
+  return (
+    <>
+      <article className="relative my-5 grid grid-cols-[1fr,min(90%,100%),1fr] gap-y-8 sm:grid-cols-[1fr,min(90%,100%),1fr] sm:pt-8 md:grid md:grid-cols-[1fr,min(80%,100%),1fr] lg:grid-cols-[1fr,min(70%,100%),1fr] xl:grid-cols-[1fr,minmax(auto,240px),min(50%,100%),minmax(auto,240px),1fr] xl:gap-x-5 xl:px-0 [&>*]:col-start-2 xl:[&>*]:col-start-3">
```
Comment (mxkaske) [inline apps/web/src/app/blog/[slug]/page.tsx:61]: You can wrap your page inside of a `Shell` component. The article doesn't have to include all the different `grid` utility classes.

File: apps/web/src/app/blog/[slug]/page.tsx [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        <div>
+          <Button variant="default">
+            <Link href="/blog">⬅ Blogs</Link>
+          </Button>
```
Comment (mxkaske) [inline apps/web/src/app/blog/[slug]/page.tsx]:
```suggestion
          <Button variant="link" asChild>
            <Link href="/blog">⬅ Blogs</Link>
          </Button>

```

File: apps/web/src/app/page.tsx [hunk 5]
```diff
@@ -37,6 +37,9 @@ export default async function Page() {
                   Star on GitHub
                 </a>
               </Button>
+              <Button variant="link">
+                <Link href="/blog">Blog</Link>
+              </Button>
```
Comment (mxkaske) [inline apps/web/src/app/page.tsx:42]:
```suggestion
              <Button variant="link" asChild>
                <Link href="/blog">Blog</Link>
              </Button>
```

File: apps/web/src/app/blog/[slug]/page.tsx (hunk 4 above)
Comment (mxkaske) [reply apps/web/src/app/blog/[slug]/page.tsx]: We have the `BackButton` component that would fit here as well!

File: apps/web/src/app/blog/[slug]/page.tsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+                {blog.publishedAtFormatted}
+              </span>
+            </div>
+            <div className="text-muted-foreground text-sm sm:pointer-events-none lg:text-base">
+              ~{blog.readingTime}
+            </div>
+          </div>
+        </section>
+
+        <main className="prose prose-p:text-sm md:prose-p:text-base prose-ul:text-sm md:prose-ul:text-base lg:prose-lg dark:prose-invert prose-code:before:content-none prose-code:after:content-none mb-6 w-full max-w-none">
```
Comment (mxkaske) [inline apps/web/src/app/blog/[slug]/page.tsx]: I'd say, just use `prose dark:prose-invert` no need to adjust the font sizes for now. Keep the html tag class names though.

File: apps/web/src/app/blog/page.tsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+export default async function Blog() {
+  const blogs = await getDisplayBlogs(allBlogs);
+
+  return (
+    <main className="mx-5 mb-7 flex flex-col items-start pt-24 sm:mx-20 md:mx-32 md:pt-28 lg:mx-60 xl:mx-96">
+      <h1 className="text-foreground font-cal mb-4 mt-2 text-5xl">Blog</h1>
+      <Button variant="secondary">
+        <Link href="/">⬅ Back to home</Link>
+      </Button>
```
Comment (mxkaske) [inline apps/web/src/app/blog/page.tsx]: Use the `BackButton` maybe?

File: apps/web/src/components/Mdx.tsx [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+  code: string;
+}
+
+export function Mdx({ code }: MdxProps) {
+  const MDXComponent = getMDXComponent(code);
+
+  return (
+    <article className="prose prose-quoteless prose-neutral dark:prose-invert max-w-none">
+      <MDXComponent components={{ ...components }} />
+    </article>
```
Comment (mxkaske) [inline apps/web/src/components/Mdx.tsx]: Here you define `prose` twice. Maybe keep that here and remove it from above.

File: apps/web/src/components/MdxComponents.tsx [hunk 9]
```diff
@@ -0,0 +1,75 @@
+import React from "react";
```
Comment (mxkaske) [inline apps/web/src/components/MdxComponents.tsx]: Let's stick with the default variables for `h1`, `h2`,... and just keep the `a` tag as it is a special case.

File: apps/web/src/components/MdxComponents.tsx [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    return (
+      <Link
+        href={href}
+        className={cn(LINK_STYLES, FOCUS_VISIBLE_OUTLINE)}
+        {...props}
+      />
+    );
+  },
```
Comment (mxkaske) [inline apps/web/src/components/MdxComponents.tsx]: Great use case!

File: apps/web/src/lib/rehypePrettyCode.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import type { Options } from "rehype-pretty-code";
```
Comment (mxkaske) [inline apps/web/src/lib/rehypePrettyCode.ts]: Can we rename the file to `rehype-pretty-code` to stick to the file name convention.

File: apps/web/src/lib/contentlayer.ts [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+        }),
+      );
+    })
+    .sort((a, b) => {
+      const [yearA, monthA, dayA] = a.publishedAt.split("-").map(Number);
+      const [yearB, monthB, dayB] = b.publishedAt.split("-").map(Number);
+      if (yearB !== yearA) return yearB - yearA;
+      if (monthB !== monthA) return monthB - monthA;
+      return dayB - dayA;
+    });
```
Comment (mxkaske) [inline apps/web/src/lib/contentlayer.ts]: We can compare the dates via timestamps like `date1.getTime() - date2.getTime() > 0`. Removes some lines.

File: apps/web/src/content/types/Blog.ts [hunk 13]
```diff
@@ -0,0 +1,58 @@
+import { defineDocumentType } from "contentlayer/source-files";
+import readingTime from "reading-time";
+
+export const Blog = defineDocumentType(() => ({
```
Comment (mxkaske) [inline apps/web/src/content/types/Blog.ts]: Let's rename to `Post` (and the corresponding names).

**Review by mxkaske — CHANGES_REQUESTED** [review]
Hey @nkilm! Thanks a lot for your PR!

Besides the comments, some structural things:

1. please use the "kebab-case" file name convention for all the files (e.g. `mdx-components.*`)
2. let's create a `@/lib/contentlayer` folder to put in the utility files that are used for it
3. The example post misses some attributes (`author` and `authorLink`)

What I am thinking for (2) is to:
- move `@/lib/contentlayer/rehype-pretty-code.ts` into it
- move the current `@/lib/contentlayer.ts` to `@/lib/contentlayer/blog.ts`
- move `@/content/types/Blog.ts` to `@/lib/contentlayer/documents/post.ts`

Regarding (3), I'd also suggest to have the possibility to have something like:

```
author:
  - name: Nkilm
  - profileImg: https://url.to/image
  - url: https://url.to/social/media
```

So just to group everything a bit more.

Good work! 🚀

Hit me up if you have questions. Also happy to pair on that one!

File: apps/web/src/app/blog/[slug]/page.tsx (hunk 2 above)
Comment (nkilm) [reply apps/web/src/app/blog/[slug]/page.tsx]: Ah I didn't notice this API, I've added the OG image now.

File: apps/web/src/app/blog/[slug]/page.tsx (hunk 4 above)
Comment (nkilm) [reply apps/web/src/app/blog/[slug]/page.tsx]: Done! I've reused `BackButton`

File: apps/web/src/app/blog/[slug]/page.tsx (hunk 6 above)
Comment (nkilm) [reply apps/web/src/app/blog/[slug]/page.tsx]: Okay, I've removed some of the classes. But the font size for code snippets look bit big😅

File: apps/web/src/components/Mdx.tsx (hunk 8 above)
Comment (nkilm) [reply apps/web/src/components/Mdx.tsx]: I didn't see that, fixed it!

File: apps/web/src/content/types/Blog.ts (hunk 13 above)
Comment (nkilm) [reply apps/web/src/content/types/Blog.ts]: Alright

File: apps/web/src/lib/contentlayer.ts (hunk 12 above)
Comment (nkilm) [reply apps/web/src/lib/contentlayer.ts]: Great!

File: apps/web/src/components/MdxComponents.tsx (hunk 9 above)
Comment (nkilm) [reply apps/web/src/components/MdxComponents.tsx]: Okay, I'll keep the default styling.

**Comment by nkilm** [comment]
@mxkaske I've implemented the requested changes, please have a look.

**Review by mxkaske — APPROVED** [review]
Thanks @nkilm for the changes! I'll merge it and will small updates here and there. 

Appreciate the work! 🚀

**Comment by nkilm** [comment]
@mxkaske Great!

**Pull request merged**